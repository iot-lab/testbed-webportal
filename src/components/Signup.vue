<template>
<div class="container mt-3">
  <div class="row">
    <div class="col-lg-8">
      <h2>Sign up to FIT IoT-LAB</h2>
      <div v-if="success" class="anlert alert-success">Success</div>
      <div v-else-if="error" class="alert alert-danger">Failed</div>
      <form v-else class=" mb-5" @submit.prevent="signup">
        <div class="text-right text-italic text-muted small">
          * All fields are mandatory
        </div>
        <div class="row">
          <div class="col-md">
            <div class="form-group">
              <label class="form-control-label">First name</label>
              <input placeholder="First name" v-model="user.firstName" class="form-control" type="text" required>
            </div>
          </div>
          <div class="col-md">                    
            <div class="form-group">
              <label class="form-control-label">Last name</label>
              <input placeholder="Last name" v-model="user.lastName" class="form-control" type="text" required>
            </div>
          </div>
        </div>
            <div class="form-group">
              <label class="form-control-label">Email</label>
              <input v-model="user.email" class="form-control" type="email" required placeholder="Academic or professional email">
              <span class="text-danger" style="font-size: 0.9rem">Please fill with an <b>academic</b> or <b>professional</b> email in order to validate your account <b>(no gmail, hotmail, &#x2026;)</b></span>
            </div>
        <div class="row">
          <div class="col-md">
            <div class="form-group">
              <label class="form-control-label">User category</label>
              <v-select v-model="user.category" :options="categories" placeholder="Category" :searchable="false" required></v-select>
            </div>                    
          </div>
          <div class="col-md">
            <div class="form-group">
              <label class="form-control-label">Organization</label>
              <input placeholder="Organization" v-model="user.organization" class="form-control" type="text" required>
            </div>
          </div>
        </div>
        <div class="row">
          <div class="col-md">
            <div class="form-group">
              <label class="form-control-label">City</label>
              <input placeholder="City" v-model="user.city" class="form-control" type="text" required>
            </div>
          </div>
          <div class="col-md">
            <div class="form-group">
              <label class="form-control-label">Country</label>
              <v-select :options="countries" v-model="user.country" required placeholder="Country" @keydown.enter.prevent=""></v-select>
            </div>
          </div>
        </div>
        <div class="form-group">
          <label class="form-control-label">Motivations</label>
          <textarea v-model="user.motivations" class="form-control" rows="5" placeholder="Tell us about you" required></textarea>
        </div>
        <div class="form-group">
          <div class="alert alert-info form-control">
            <i class="fa fa-pencil"></i> <b>Tell us about your motivations</b>
            <ul style="margin: 0">
              <li>Research domain (Radio communication, networking protocol, distributed applications, &#x2026;)</li>
              <li>What kind of experiments do you want to run with IoT-LAB?</li>
              <li>Goal (Verify something existing in large scale, new development, &#x2026;)</li>
              <li>Experience with sensor networks (n00b, experiments with X platform, former IoT-LAB user, Guru, God)</li>
            </ul>
          </div>
        </div>
        <div class="form-group">
          <vue-recaptcha ref="recaptcha" @verify="onCaptchaVerify" @expired="onCaptchaExpired" :sitekey="reCaptchaSitekey"></vue-recaptcha>
        </div>
        <div class="form-check">
          <label class="custom-control custom-checkbox">
            <input v-model="charter" type="checkbox" class="custom-control-input">
            <span class="custom-control-indicator"></span>
            <span class="custom-control-description">
              I read and I accept <a href="/charter/" target="_blank">IoT-LAB Terms of Service</a>.
            </span>
          </label>
        </div>
        <div class="form-group">
          <button class="btn btn-lg btn-primary" type="submit">Sign up</button>
          <router-link to="login" class="btn btn-lg btn-secondary">Login</router-link>
        </div>
      
      </form>

    </div>
  </div> <!-- row -->
</div>
</template>

<script>
import vSelect from 'vue-select'
import VueRecaptcha from 'vue-recaptcha'
import {iotlab} from '@/rest'
import countries from '@/assets/js/countries'
import categories from '@/assets/js/categories'

export default {
  name: 'signup',
  components: {vSelect, VueRecaptcha},

  data () {
    return {
      reCaptchaSitekey: '6Ld8cR4UAAAAAC-zBLP9m2bC35xyyYwTbvkBcx4q',
      charter: false,
      verified: false,
      user: {
        'sshPublicKeys': [''],
      },
      countries: countries,
      categories: categories,
      category: undefined,
    }
  },

  methods: {
    async signup () {
      // check charter read
      if (!this.charter) {
        alert('Please read and accept the Terms of Service')
        return
      }
      // check reCaptcha verified
      if (!this.verified) {
        alert('Please verify the captcha "I\'m not a robot"')
      }

      try {
        await iotlab.signup(this.user)
        // this.$router.push('dashboard')
        alert('Success')
        this.success = true
      } catch (err) {
        this.error = true
        alert('Error')
      }
    },
    onCaptchaVerify: function (response) {
      console.log('Captcha Verified')
      this.verified = true
    },
    onCaptchaExpired: function () {
      console.log('Captcha Expired')
      this.verified = false
    },
  },
}
</script>

<style scoped>
ul {
  padding-left: 15px;
}
.v-select {
  background: white;
}
</style>
