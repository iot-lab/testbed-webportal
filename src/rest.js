import axios from 'axios'

const API_URL = `https://${process.env.VUE_APP_IOTLAB_HOST}/api/`

export const iotlab = {

  api: axios.create({baseURL: API_URL, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')}),

  create (username, password) {
    this.api = axios.create({
      baseURL: API_URL,
      auth: {
        username: username,
        password: password,
      },
    })
  },

  // USERS API

  async signup (user, nbUsers) {
    if (nbUsers) {
      // create multiple user accounts
      await iotlab.api.post(`/users?nbusers=${nbUsers}`, user)
    } else {
      // create single user account
      await iotlab.api.post('/users', user)
    }
  },

  async activateAccount (hash) {
    await iotlab.api.post('/users/activate/', {'hash': hash})
  },

  async deleteUser (login = undefined) {
    if (login) {
      await iotlab.api.delete(`/users/${login}`)
    } else {
      await iotlab.api.delete(`/user`)
    }
  },

  async resetPassword (email) {
    await iotlab.api.post(`/users/reset_password`, {'email': email})
  },

  async changePassword (oldPwd, newPwd, confirmPwd) {
    await iotlab.api.put('/user/password', {
      'old_password': oldPwd,
      'new_password': newPwd,
      'confirm_new_password': confirmPwd,
    })
  },

  async getSSHkeys () {
    return iotlab.api.get('/user/keys').then(resp => resp.data.sshkeys)
  },

  async modifySSHkeys (keys) {
    return iotlab.api.post('/user/keys', {'sshkeys': keys})
  },

  async deleteSSHkey (id) {
    return iotlab.api.delete(`/user/keys/${id}`)
  },

  async getUsers ({status, group, search} = {}) {
    let params = {}
    if (status) {
      params.status = status
    }
    if (group) {
      params.group = group
    }
    if (search) {
      params.search = search
    }
    return iotlab.api.get('/users', {params: params}).then(resp => resp.data)
  },

  async getUserInfo (login = undefined) {
    if (login) {
      return iotlab.api.get(`/users/${login}`).then(resp => resp.data)
    } else {
      return iotlab.api.get('/user').then(resp => resp.data)
    }
  },

  async setUserInfo (user, login = undefined) {
    if (login) {
      return iotlab.api.put(`/users/${login}`, user)
    } else {
      return iotlab.api.put('/user', user)
    }
  },

  async activateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'active'})
  },

  async deactivateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'inactive'})
  },

  async setUserGroups (login, groups) {
    return iotlab.api.put(`/users/${login}`, {'groups': groups})
  },

  async sendMail (recipients, subject, message) {
    await iotlab.api.post('/users/email', {
      recipients: recipients,
      subject: subject,
      body: message,
    })
  },

  // EXPERIMENT API

  async getUserExperimentsCount () {
    return iotlab.api.get('/experiments/total').then(resp => resp.data)
  },

  async getAllExperimentsCount (user) {
    if (user) {
      return iotlab.api.get(`/experiments/total/users/${user}`).then(resp => resp.data)
    } else {
      return iotlab.api.get(`/experiments/total/all`).then(resp => resp.data)
    }
  },

  async getAllExperiments ({limit, offset, user, state = 'Terminated,Error,Running,Finishing,Resuming,toError,Waiting,Launching,Hold,toLaunch,toAckReservation,Suspended'} = {}) {
    let params = {}
    if (state) {
      params.state = state
    }
    if (limit) {
      params.limit = Math.min(limit, 500) // IoT-LAB api restrict queries to 500 entries
    }
    if (offset) {
      params.offset = offset
    }
    if (user === '@self') {
      return iotlab.api.get('/experiments', {params: params}).then(resp => resp.data.items)
    } else if (user) {
      return iotlab.api.get(`/experiments/users/${user}`, {params: params}).then(resp => resp.data.items)
    } else {
      return iotlab.api.get('/experiments/all', {params: params}).then(resp => resp.data.items)
    }
  },

  async getRunningExperiments () {
    return iotlab.api.get('/experiments/running').then(resp => resp.data.items)
  },

  async getExperiment (id) {
    return iotlab.api.get(`/experiments/${id}`).then(resp => resp.data)
  },

  async getExperimentArchive (id) {
    return iotlab.api.get(`/experiments/${id}/data`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async getExperimentDeployment (id) {
    return iotlab.api.get(`/experiments/${id}/deployment`).then(resp => resp.data)
  },

  async stopExperiment (id) {
    return iotlab.api.delete(`/experiments/${id}`)
  },

  async submitExperiment ({type, name, duration, reservation, nodes, firmwareassociations, firmwares, profileassociations, profiles} = {}) {
    const formData = new FormData()
    formData.append('experiment.json', JSON.stringify({
      type: type,
      duration: duration,
      reservation: reservation,
      name: name,
      nodes: nodes,
      firmwareassociations: firmwareassociations,
      profileassociations: profileassociations,
    }))
    // attached firmware names should be unique
    let attachedFirmwares = []
    for (let asso of firmwareassociations || []) {
      if (asso.firmwarename in firmwares && !attachedFirmwares.includes(asso.firmwarename)) {
        formData.append(asso.firmwarename, new Blob([firmwares[asso.firmwarename]], {type: 'application/octet-stream'}), asso.firmwarename)
        attachedFirmwares.push(asso.firmwarename)
      }
    }
    return iotlab.api.post('/experiments', formData).then(resp => resp.data)
  },

  async checkFirmware (fileString) {
    const formData = new FormData()
    formData.append('firmware.bin', fileString)
    return iotlab.api.post('/firmwares/checker', formData).then(resp => resp.data)
  },

  async flashFirmware (id, nodes, fileString) {
    const formData = new FormData()
    formData.append(`experiment${id}.json`, JSON.stringify(nodes))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/experiments/${id}/nodes/flash`, formData).then(resp => resp.data)
  },

  async flashResourcesFirmware (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/flash/${name}`, nodes).then(resp => resp.data)
  },

  async updateMonitoring (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/monitoring/${name}`, nodes).then(resp => resp.data)
  },

  async updateExperimentMobilityCircuit (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/robots/mobility/${name}`, nodes).then(resp => resp.data)
  },

  async reloadExperiment (id) {
    return iotlab.api.post(`/experiments/${id}/reload`, {}).then(resp => resp.data)
  },

  // USER GROUPS

  async getUserGroups () {
    return iotlab.api.get('/groups').then(resp => resp.data)
  },

  async createGroup (name) {
    return iotlab.api.post(`/groups`, {name: name})
  },

  async deleteGroup (group) {
    return iotlab.api.delete(`/groups/${group}`)
  },

  // OTHER API

  async getSites () {
    return iotlab.api.get('/sites').then(resp => resp.data.items)
  },

  async getSitesDetails () {
    return iotlab.api.get('/sites/details').then(resp => resp.data.items)
  },

  async getNodes () {
    return iotlab.api.get('/nodes').then(resp => resp.data.items)
  },

  async updateNodesProperties () {
    return iotlab.api.put('/nodes')
  },

  async updateNodesPropertiesOldApi () {
    const OLD_API = `https://${process.env.VUE_APP_IOTLAB_HOST}/rest/`
    const oldApi = axios.create({baseURL: OLD_API, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')})
    return oldApi.post('/admin/nodes')
  },

  async getNodesIds () {
    return iotlab.api.get('/nodes/ids').then(resp => resp.data.items)
  },

  async getExperimentNodes (id) {
    return iotlab.api.get(`/experiments/${id}/nodes`).then(resp => resp.data.items)
  },

  async getExperimentToken (id) {
    return iotlab.api.get(`/experiments/${id}/token`).then(resp => resp.data.token)
  },

  async sendNodesCommand (id, cmd, nodes) {
    // cmd in [start, stop, reset, flash-idle, profile-reset, debug-start, debug-stop]
    return iotlab.api.post(`/experiments/${id}/nodes/${cmd}`, nodes).then(resp => resp.data)
  },

  // STORES (firmwares, profiles, etc)

  async getMobilityCircuits () {
    return iotlab.api.get('/mobilities/circuits').then(resp => resp.data.items)
  },

  async getMobilityCircuit (name) {
    return iotlab.api.get(`/mobilities/circuits/${name}`).then(resp => resp.data)
  },

  async getRobotsSiteMapImage (site) {
    return iotlab.api.get(`/robots/${site}/map/image`, {
      responseType: 'arraybuffer',
      headers: {
        'Access-Control-Allow-Origin': '*',
      },
    }).then(resp => Buffer.from(resp.data, 'binary').toString('base64')
    )
  },

  async getRobotsSiteMapConfig (site) {
    return iotlab.api.get(`/robots/${site}/map/config`).then(resp => resp.data)
  },

  async createMobilityCircuit (circuit) {
    return iotlab.api.post(`/mobilities/circuits`, circuit)
  },

  async updateMobilityCircuit (name, circuit) {
    return iotlab.api.put(`/mobilities/circuits/${name}`, circuit)
  },

  async deleteMobilityCircuit (name) {
    return iotlab.api.delete(`/mobilities/circuits/${name}`)
  },

  async getMonitoringProfiles () {
    return iotlab.api.get('/monitoring').then(resp => resp.data)
  },

  async getMonitoringProfile (name) {
    return iotlab.api.get(`/monitoring/${name}`).then(resp => resp.data)
  },

  async updateMonitoringProfile (name, profile) {
    return iotlab.api.put(`/monitoring/${name}`, profile)
  },

  async createMonitoringProfile (profile) {
    return iotlab.api.post(`/monitoring`, profile)
  },

  async deleteMonitoringProfile (name) {
    return iotlab.api.delete(`/monitoring/${name}`)
  },

  async getFirmwares () {
    return iotlab.api.get('/firmwares').then(resp => resp.data)
  },

  async getFirmware (name) {
    return iotlab.api.get(`/firmwares/${name}`).then(resp => resp.data)
  },

  async getFirmwareFile (name) {
    return iotlab.api.get(`/firmwares/${name}/file`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async updateFirmware (name, firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    if (fileString) {
      formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    }
    return iotlab.api.put(`/firmwares/${name}`, formData).then(resp => resp.data)
  },

  async createFirmware (firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/firmwares`, formData).then(resp => resp.data)
  },

  async deleteFirmware (name) {
    return iotlab.api.delete(`/firmwares/${name}`)
  },
}
