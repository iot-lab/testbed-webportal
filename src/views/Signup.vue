<template>
<div class="container mt-3">
  <div class="row justify-content-center">
    <div class="col-lg-8">
      <h2>Sign up to FIT IoT-LAB</h2>
      <div class="card border-danger bg-danger my-2 mx-auto" v-if="hostname === 'devwww.iot-lab.info'">
        <div class="card-body text-white p-2">
          Signup on <b>devwww</b> is restricted to <b>administrators</b>.
          Users should sign up to <a class="text-white alert-link" href="https://www.iot-lab.info/testbed">www.iot-lab.info</a>.
        </div>
      </div>
      <div v-if="success" class="card border-success my-5">
        <div class="card-header bg-success text-white"><i class="fa fa-check mr-1"></i> An email has been sent to <i>{{user.email}}</i></div>
        <div class="card-body text-success">
          <h5>Check your inbox and follow the link to activate your account.</h5>
        </div>
      </div>
      <form v-else class="mb-5" @submit.prevent="signup">
        <div class="text-right text-italic text-muted small">
          * All fields are mandatory
        </div>

        <user-form :user="user" ref="user" :hidden="['groups']"></user-form>

        <div class="form-group">
          <vue-recaptcha ref="recaptcha" @verify="onCaptchaVerify" @expired="onCaptchaExpired"
            :sitekey="reCaptchaSitekey"
            :class="(!captcha.verified && captcha.dirty) ? 'border-danger form-control p-2' : ''">
          </vue-recaptcha>
          <div class="invalid-feedback mb-3" :class="{'d-block': !captcha.verified && captcha.dirty }" v-show="!captcha.verified && captcha.dirty">
            Please verify the captcha.
          </div>
        </div>
        <div class="form-check">
          <label class="custom-control custom-control-inline custom-checkbox">
            <input v-model="charter" name="charter" type="checkbox" class="custom-control-input" v-validate="'required'" :class="{'is-invalid': errors.has('charter') }">
            <span class="custom-control-label">
              I read and I accept <a href="/charter/" :class="{'text-danger': errors.has('charter') }" target="_blank">IoT-LAB Terms of Service</a>.
            </span>
          </label>
          <div class="invalid-feedback mb-3" :class="{'d-block': errors.has('charter') }" v-show="errors.has('charter')">
            You must check the box and accept the terms of service.
          </div>
        </div>
        <div class="form-group mt-3">
          <button class="btn btn-lg btn-primary mr-1" type="submit">Sign up</button>
          <router-link :to="{name:'login'}" class="btn btn-lg btn-secondary">Login</router-link>
        </div>
      </form>

    </div>
  </div> <!-- row -->
</div>
</template>

<script>
import VueRecaptcha from 'vue-recaptcha'
import {iotlab} from '@/rest'
import UserForm from '@/components/UserForm'

export default {
  name: 'signup',
  components: {UserForm, VueRecaptcha},

  data () {
    return {
      hostname: location.hostname,
      reCaptchaSitekey: '6Ld8cR4UAAAAAC-zBLP9m2bC35xyyYwTbvkBcx4q',
      user: {
        'sshkeys': [''],
      },
      charter: false,
      success: false,
      captcha: {
        dirty: false,
        verified: false,
      },
    }
  },

  methods: {
    signup () {
      this.$validator.validateAll().then(async (validated) => {
        if (!(await this.$refs.user.validate()) || !validated || !this.captcha.verified) {
          this.captcha.dirty = true
          console.log('Form is not valid.')
          return
        }
        if (location.hostname === 'devwww.iot-lab.info') {
          alert('Signup on devwww is restricted to IoT-LAB administrators')
          return
        }
        try {
          await iotlab.signup(this.user)
          this.success = true
        } catch (err) {
          this.success = false
          this.$notify({text: 'Error during signup<br>' + err.response.data.message, type: 'error', duration: -1})
        }
      })
    },
    onCaptchaVerify: function (response) {
      console.log('Captcha Verified')
      this.captcha.verified = true
    },
    onCaptchaExpired: function () {
      console.log('Captcha Expired')
      this.captcha.verified = false
    },
  },
}
</script>
