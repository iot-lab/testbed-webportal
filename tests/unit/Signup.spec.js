import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import Signup from '@/views/Signup'
import { iotlab } from '@/rest'
jest.mock('@/rest')

// Cannot test using VeeValidate due to a bug, should try different lib versions or wait for a fix
// https://github.com/vuejs/vue-test-utils/issues/829
//
// workaround:
// localVue.use(VeeValidate)
const errors = {
  has: jest.fn(),
  first: jest.fn(),
}

const localVue = createLocalVue()
localVue.use(VueRouter)
const router = new VueRouter()

describe('Signup.vue', () => {

  it('should match snapshot', () => {
    const wrapper = shallowMount(Signup, {localVue, router, mocks: {errors}})
    expect(wrapper.element).toBeDefined()
    expect(wrapper.element).toMatchSnapshot()
  })

})
