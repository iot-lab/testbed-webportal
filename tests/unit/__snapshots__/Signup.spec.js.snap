// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup.vue should match snapshot 1`] = `
<div
  class="container mt-3"
>
  <div
    class="row justify-content-center"
  >
    <div
      class="col-lg-8"
    >
      <h2>
        Sign up to FIT IoT-LAB
      </h2>
       
      <!---->
       
      <form
        class="mb-5"
      >
        <div
          class="text-right text-italic text-muted small"
        >
          
          * All fields are mandatory
        
        </div>
         
        <userform-stub
          hidden="groups"
          mode="create"
          user="[object Object]"
        />
         
        <div
          class="form-group"
        >
          <vuerecaptcha-stub
            class=""
            sitekey="6Ld8cR4UAAAAAC-zBLP9m2bC35xyyYwTbvkBcx4q"
          />
           
          <div
            class="invalid-feedback mb-3"
            style="display: none;"
          >
            
            Please verify the captcha.
          
          </div>
        </div>
         
        <div
          class="form-check"
        >
          <label
            class="custom-control custom-checkbox"
          >
            <input
              class="custom-control-input"
              name="charter"
              type="checkbox"
            />
             
            <span
              class="custom-control-indicator"
            />
             
            <span
              class="custom-control-description"
            >
              
              I read and I accept 
              <a
                class=""
                href="/charter/"
                target="_blank"
              >
                IoT-LAB Terms of Service
              </a>
              .
            
            </span>
          </label>
           
          <div
            class="invalid-feedback mb-3"
            style="display: none;"
          >
            
            You must check the box and accept the terms of service.
          
          </div>
        </div>
         
        <div
          class="form-group"
        >
          <button
            class="btn btn-lg btn-primary mr-1"
            type="submit"
          >
            Sign up
          </button>
           
          <router-link-stub
            class="btn btn-lg btn-secondary"
            event="click"
            tag="a"
            to="[object Object]"
          >
            Login
          </router-link-stub>
        </div>
      </form>
    </div>
  </div>
</div>
`;
